File: apps/php-sdk/src/Client/FirecrawlClient.php
<?php

declare(strict_types=1);

namespace Firecrawl\Client;

use Firecrawl\Exceptions\FirecrawlException;
use Firecrawl\Exceptions\JobTimeoutException;
use Firecrawl\Models\AgentOptions;
use Firecrawl\Models\AgentResponse;
use Firecrawl\Models\AgentStatusResponse;
use Firecrawl\Models\BatchScrapeJob;
use Firecrawl\Models\BatchScrapeOptions;
use Firecrawl\Models\BatchScrapeResponse;
use Firecrawl\Models\BrowserCreateResponse;
use Firecrawl\Models\BrowserDeleteResponse;
use Firecrawl\Models\BrowserExecuteResponse;
use Firecrawl\Models\BrowserListResponse;
use Firecrawl\Models\ConcurrencyCheck;
use Firecrawl\Models\CrawlJob;
use Firecrawl\Models\CrawlOptions;
use Firecrawl\Models\CrawlResponse;
use Firecrawl\Models\CreditUsage;
use Firecrawl\Models\Document;
use Firecrawl\Models\MapData;
use Firecrawl\Models\MapOptions;
use Firecrawl\Models\ScrapeOptions;
use Firecrawl\Models\SearchData;
use Firecrawl\Models\SearchOptions;
use GuzzleHttp\ClientInterface;

final class FirecrawlClient
{
    private const DEFAULT_API_URL = 'https://api.firecrawl.dev';
    private const DEFAULT_TIMEOUT_SECONDS = 300;
    private const DEFAULT_MAX_RETRIES = 3;
    private const DEFAULT_BACKOFF_FACTOR = 0.5;
    private const DEFAULT_POLL_INTERVAL = 2;
    private const DEFAULT_JOB_TIMEOUT = 300;

    private readonly FirecrawlHttpClient $http;

    private function __construct(FirecrawlHttpClient $http)
    {
        $this->http = $http;
    }

    /**
     * Create a client with named parameters.
     *
     * Uses FIRECRAWL_API_KEY and FIRECRAWL_API_URL environment variables as fallbacks.
     */
    public static function create(
        ?string $apiKey = null,
        ?string $apiUrl = null,
        float $timeoutSeconds = self::DEFAULT_TIMEOUT_SECONDS,
        int $maxRetries = self::DEFAULT_MAX_RETRIES,
        float $backoffFactor = self::DEFAULT_BACKOFF_FACTOR,
        ?ClientInterface $httpClient = null,
    ): self {
        $resolvedKey = trim($apiKey ?: (getenv('FIRECRAWL_API_KEY') ?: ''));
        if ($resolvedKey === '') {
            throw new FirecrawlException(
                'API key is required. Pass it directly or set the FIRECRAWL_API_KEY environment variable.',
            );
        }

        $resolvedUrl = $apiUrl ?: (getenv('FIRECRAWL_API_URL') ?: self::DEFAULT_API_URL);

        if (!preg_match('#^https?://#i', $resolvedUrl)) {
            throw new FirecrawlException(
                'API URL must be a fully qualified URL including scheme (e.g. https://api.firecrawl.dev).',
            );
        }

        $http = new FirecrawlHttpClient(
            $resolvedKey,
            $resolvedUrl,
            $timeoutSeconds,
            $maxRetries,
            $backoffFactor,
            $httpClient,
        );

        return new self($http);
    }

    /**
     * Create a client from the FIRECRAWL_API_KEY environment variable.
     */
    public static function fromEnv(): self
    {
        return self::create();
    }

    // ================================================================
    // SCRAPE
    // ================================================================

    /**
     * Scrape a single URL and return the document.
     */
    public function scrape(string $url, ?ScrapeOptions $options = null): Document
    {
        $body = ['url' => $url];
        if ($options !== null) {
            $body = array_merge($body, $options->toArray());
        }

        $response = $this->http->post('/v2/scrape', $body);

        return Document::fromArray($response['data'] ?? $response);
    }

    /**
     * Interact with the scrape-bound browser session for a scrape job.
     */
    public function interact(
        string $jobId,
        string $code,
        string $language = 'node',
        ?int $timeout = null,
        ?string $origin = null,
    ): BrowserExecuteResponse {
        $body = [
            'code' => $code,
            'language' => $language,
        ];
        if ($timeout !== null) {
            $body['timeout'] = $timeout;
        }
        if ($origin !== null) {
            $body['origin'] = $origin;
        }

        return BrowserExecuteResponse::fromArray(
            $this->http->post("/v2/scrape/{$jobId}/interact", $body),
        );
    }

    /**
     * Stop the interactive browser session for a scrape job.
     */
    public function stopInteractiveBrowser(string $jobId): BrowserDeleteResponse
    {
        return BrowserDeleteResponse::fromArray(
            $this->http->delete("/v2/scrape/{$jobId}/interact"),
        );
    }

    // ================================================================
    // CRAWL
    // ================================================================

    /**
     * Start an async crawl job and return immediately.
     */
    public function startCrawl(string $url, ?CrawlOptions $options = null): CrawlResponse
    {
        $body = ['url' => $url];
        if ($options !== null) {
            $body = array_merge($body, $options->toArray());
        }

        return CrawlResponse::fromArray($this->http->post('/v2/crawl', $body));
    }

    /**
     * Get the status and results of a crawl job.
     */
    public function getCrawlStatus(string $jobId): CrawlJob
    {
        return CrawlJob::fromArray($this->http->get("/v2/crawl/{$jobId}"));
    }

    /**
     * Crawl a website and wait for completion (auto-polling).
     */
    public function crawl(
        string $url,
        ?CrawlOptions $options = null,
        int $pollIntervalSec = self::DEFAULT_POLL_INTERVAL,
        int $timeoutSec = self::DEFAULT_JOB_TIMEOUT,
    ): CrawlJob {
        $start = $this->startCrawl($url, $options);

        return $this->pollCrawl($start->getId(), $pollIntervalSec, $timeoutSec);
    }

    /**
     * Cancel a running crawl job.
     *
     * @return array<string, mixed>
     */
    public function cancelCrawl(string $jobId): array
    {
        return $this->http->delete("/v2/crawl/{$jobId}");
    }

    /**
     * Get errors from a crawl job.
     *
     * @return array<string, mixed>
     */
    public function getCrawlErrors(string $jobId): array
    {
        return $this->http->get("/v2/crawl/{$jobId}/errors");
    }

    // ================================================================
    // BATCH SCRAPE
    // ================================================================

    /**
     * Start an async batch scrape job.
     *
     * @param list<string> $urls
     */
    public function startBatchScrape(array $urls, ?BatchScrapeOptions $options = null): BatchScrapeResponse
    {
        $body = ['urls' => $urls];
        $extraHeaders = [];

        if ($options !== null) {
            $idempotencyKey = $options->getIdempotencyKey();
            if ($idempotencyKey !== null && $idempotencyKey !== '') {
                $extraHeaders['x-idempotency-key'] = $idempotencyKey;
            }

            $body = array_merge($body, $options->toArray());
        }

        return BatchScrapeResponse::fromArray(
            $this->http->post('/v2/batch/scrape', $body, $extraHeaders),
        );
    }

    /**
     * Get the status and results of a batch scrape job.
     */
    public function getBatchScrapeStatus(string $jobId): BatchScrapeJob
    {
        return BatchScrapeJob::fromArray($this->http->get("/v2/batch/scrape/{$jobId}"));
    }

    /**
     * Batch-scrape URLs and wait for completion (auto-polling).
     *
     * @param list<string> $urls
     */
    public function batchScrape(
        array $urls,
        ?BatchScrapeOptions $options = null,
        int $pollIntervalSec = self::DEFAULT_POLL_INTERVAL,
        int $timeoutSec = self::DEFAULT_JOB_TIMEOUT,
    ): BatchScrapeJob {
        $start = $this->startBatchScrape($urls, $options);

        return $this->pollBatchScrape($start->getId(), $pollIntervalSec, $timeoutSec);
    }

    /**
     * Cancel a running batch scrape job.
     *
     * @return array<string, mixed>
     */
    public function cancelBatchScrape(string $jobId): array
    {
        return $this->http->delete("/v2/batch/scrape/{$jobId}");
    }

    // ================================================================
    // MAP
    // ================================================================

    /**
     * Discover URLs on a website.
     */
    public function map(string $url, ?MapOptions $options = null): MapData
    {
        $body = ['url' => $url];
        if ($options !== null) {
            $body = array_merge($body, $options->toArray());
        }

        $response = $this->http->post('/v2/map', $body);

        return MapData::fromArray($response['data'] ?? $response);
    }

    // ================================================================
    // SEARCH
    // ================================================================

    /**
     * Perform a web search.
     */
    public function search(string $query, ?SearchOptions $options = null): SearchData
    {
        $body = ['query' => $query];
        if ($options !== null) {
            $body = array_merge($body, $options->toArray());
        }

        $response = $this->http->post('/v2/search', $body);

        return SearchData::fromArray($response['data'] ?? $response);
    }

    // ================================================================
    // AGENT
    // ================================================================

    /**
     * Start an async agent task.
     */
    public function startAgent(AgentOptions $options): AgentResponse
    {
        return AgentResponse::fromArray(
            $this->http->post('/v2/agent', $options->toArray()),
        );
    }

    /**
     * Get the status of an agent task.
     */
    public function getAgentStatus(string $jobId): AgentStatusResponse
    {
        return AgentStatusResponse::fromArray(
            $this->http->get("/v2/agent/{$jobId}"),
        );
    }

    /**
     * Run an agent task and wait for completion (auto-polling).
     */
    public function agent(
        AgentOptions $options,
        int $pollIntervalSec = self::DEFAULT_POLL_INTERVAL,
        int $timeoutSec = self::DEFAULT_JOB_TIMEOUT,
    ): AgentStatusResponse {
        $start = $this->startAgent($options);

        if ($start->getId() === null) {
            throw new FirecrawlException('Agent start did not return a job ID');
        }

        $this->ensureValidPollInterval($pollIntervalSec);

        $deadline = time() + $timeoutSec;
        while (time() < $deadline) {
            $status = $this->getAgentStatus($start->getId());
            if ($status->isDone()) {
                return $status;
            }
            sleep($pollIntervalSec);
        }

        throw new JobTimeoutException($start->getId(), $timeoutSec, 'Agent');
    }

    /**
     * Cancel a running agent task.
     *
     * @return array<string, mixed>
     */
    public function cancelAgent(string $jobId): array
    {
        return $this->http->delete("/v2/agent/{$jobId}");
    }

    // ================================================================
    // BROWSER
    // ================================================================

    /**
     * Create a new browser session.
     */
    public function browser(
        ?int $ttl = null,
        ?int $activityTtl = null,
        ?bool $streamWebView = null,
    ): BrowserCreateResponse {
        $body = [];
        if ($ttl !== null) {
            $body['ttl'] = $ttl;
        }
        if ($activityTtl !== null) {
            $body['activityTtl'] = $activityTtl;
        }
        if ($streamWebView !== null) {
            $body['streamWebView'] = $streamWebView;
        }

        return BrowserCreateResponse::fromArray($this->http->post('/v2/browser', $body));
    }

    /**
     * Execute code in a browser session.
     */
    public function browserExecute(
        string $sessionId,
        string $code,
        string $language = 'bash',
        ?int $timeout = null,
    ): BrowserExecuteResponse {
        $body = [
            'code' => $code,
            'language' => $language,
        ];
        if ($timeout !== null) {
            $body['timeout'] = $timeout;
        }

        return BrowserExecuteResponse::fromArray(
            $this->http->post("/v2/browser/{$sessionId}/execute", $body),
        );
    }

    /**
     * Delete a browser session.
     */
    public function deleteBrowser(string $sessionId): BrowserDeleteResponse
    {
        return BrowserDeleteResponse::fromArray(
            $this->http->delete("/v2/browser/{$sessionId}"),
        );
    }

    /**
     * List browser sessions.
     */
    public function listBrowsers(?string $status = null): BrowserListResponse
    {
        $endpoint = '/v2/browser';
        if ($status !== null && $status !== '') {
            $endpoint .= '?status=' . urlencode($status);
        }

        return BrowserListResponse::fromArray($this->http->get($endpoint));
    }

    // ================================================================
    // USAGE & METRICS
    // ================================================================

    /**
     * Get current concurrency usage.
     */
    public function getConcurrency(): ConcurrencyCheck
    {
        return ConcurrencyCheck::fromArray($this->http->get('/v2/concurrency-check'));
    }

    /**
     * Get current credit usage.
     */
    public function getCreditUsage(): CreditUsage
    {
        return CreditUsage::fromArray($this->http->get('/v2/team/credit-usage'));
    }

    // ================================================================
    // INTERNAL POLLING HELPERS
    // ================================================================

    private function ensureValidPollInterval(int $pollIntervalSec): void
    {
        if ($pollIntervalSec < 1) {
            throw new FirecrawlException('Poll interval must be at least 1 second, got ' . $pollIntervalSec);
        }
    }

    private function pollCrawl(
        ?string $jobId,
        int $pollIntervalSec,
        int $timeoutSec,
    ): CrawlJob {
        if ($jobId === null) {
            throw new FirecrawlException('Crawl start did not return a job ID');
        }

        $this->ensureValidPollInterval($pollIntervalSec);

        $deadline = time() + $timeoutSec;
        while (time() < $deadline) {
            $job = $this->getCrawlStatus($jobId);
            if ($job->isDone()) {
                return $this->paginateCrawl($job);
            }
            sleep($pollIntervalSec);
        }

        throw new JobTimeoutException($jobId, $timeoutSec, 'Crawl');
    }

    private function pollBatchScrape(
        ?string $jobId,
        int $pollIntervalSec,
        int $timeoutSec,
    ): BatchScrapeJob {
        if ($jobId === null) {
            throw new FirecrawlException('Batch scrape start did not return a job ID');
        }

        $this->ensureValidPollInterval($pollIntervalSec);

        $deadline = time() + $timeoutSec;
        while (time() < $deadline) {
            $job = $this->getBatchScrapeStatus($jobId);
            if ($job->isDone()) {
                return $this->paginateBatchScrape($job);
            }
            sleep($pollIntervalSec);
        }

        throw new JobTimeoutException($jobId, $timeoutSec, 'Batch scrape');
    }

    private function assertSameOrigin(string $url): void
    {
        $baseHost = parse_url($this->http->getBaseUrl(), PHP_URL_HOST);
        $nextHost = parse_url($url, PHP_URL_HOST);

        if ($baseHost === null || $nextHost === null || strcasecmp($baseHost, $nextHost) !== 0) {
            throw new FirecrawlException(
                'Pagination URL origin does not match the API base URL. Refusing to follow: ' . $url,
            );
        }
    }

    private function paginateCrawl(CrawlJob $job): CrawlJob
    {
        $current = $job;
        while ($current->getNext() !== null && $current->getNext() !== '') {
            $this->assertSameOrigin($current->getNext());
            $nextRaw = $this->http->getAbsolute($current->getNext());
            $nextPage = CrawlJob::fromArray($nextRaw);

            foreach ($nextPage->getData() as $doc) {
                $job->appendData($doc);
            }

            $current = $nextPage;
        }

        return $job;
    }

    private function paginateBatchScrape(BatchScrapeJob $job): BatchScrapeJob
    {
        $current = $job;
        while ($current->getNext() !== null && $current->getNext() !== '') {
            $this->assertSameOrigin($current->getNext());
            $nextRaw = $this->http->getAbsolute($current->getNext());
            $nextPage = BatchScrapeJob::fromArray($nextRaw);

            foreach ($nextPage->getData() as $doc) {
                $job->appendData($doc);
            }

            $current = $nextPage;
        }

        return $job;
    }
}

File: apps/.net-sdk/Firecrawl/FirecrawlClient.cs
using System.Text.Json;
using Firecrawl.Exceptions;
using Firecrawl.Models;

namespace Firecrawl;

/// <summary>
/// Client for the Firecrawl v2 API.
///
/// <example>
/// <code>
/// var client = new FirecrawlClient("fc-your-api-key");
///
/// // Scrape a single page
/// var doc = await client.ScrapeAsync("https://example.com",
///     new ScrapeOptions { Formats = new List&lt;object&gt; { "markdown" } });
///
/// // Crawl a website
/// var job = await client.CrawlAsync("https://example.com",
///     new CrawlOptions { Limit = 50 });
/// </code>
/// </example>
/// </summary>
public class FirecrawlClient
{
    private const string DefaultApiUrl = "https://api.firecrawl.dev";
    private static readonly TimeSpan DefaultTimeout = TimeSpan.FromMinutes(5);
    private const int DefaultMaxRetries = 3;
    private const double DefaultBackoffFactor = 0.5;
    private const int DefaultPollIntervalSec = 2;
    private const int DefaultJobTimeoutSec = 300;

    private readonly FirecrawlHttpClient _http;

    /// <summary>
    /// Creates a new FirecrawlClient with the specified API key.
    /// </summary>
    /// <param name="apiKey">The Firecrawl API key.</param>
    /// <param name="apiUrl">Optional API base URL (defaults to https://api.firecrawl.dev).</param>
    /// <param name="timeout">Optional HTTP request timeout.</param>
    /// <param name="maxRetries">Optional maximum number of retries for transient failures.</param>
    /// <param name="backoffFactor">Optional exponential backoff factor in seconds.</param>
    /// <param name="httpClient">Optional pre-configured HttpClient instance.</param>
    public FirecrawlClient(
        string? apiKey = null,
        string? apiUrl = null,
        TimeSpan? timeout = null,
        int maxRetries = DefaultMaxRetries,
        double backoffFactor = DefaultBackoffFactor,
        HttpClient? httpClient = null)
    {
        var resolvedKey = ResolveApiKey(apiKey);
        var resolvedUrl = ResolveApiUrl(apiUrl);

        _http = new FirecrawlHttpClient(
            resolvedKey,
            resolvedUrl,
            timeout ?? DefaultTimeout,
            maxRetries,
            backoffFactor,
            httpClient);
    }

    // ================================================================
    // SCRAPE
    // ================================================================

    /// <summary>
    /// Scrapes a single URL and returns the document.
    /// </summary>
    public async Task<Document> ScrapeAsync(
        string url,
        ScrapeOptions? options = null,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(url);

        var body = BuildBody(options);
        body["url"] = url;

        var response = await _http.PostAsync<ApiResponse<Document>>(
            "/v2/scrape", body, cancellationToken: cancellationToken);

        return response.Data ?? throw new FirecrawlException("Scrape response contained no data");
    }

    // ================================================================
    // CRAWL
    // ================================================================

    /// <summary>
    /// Starts an async crawl job and returns immediately.
    /// </summary>
    public async Task<CrawlResponse> StartCrawlAsync(
        string url,
        CrawlOptions? options = null,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(url);

        var body = BuildBody(options);
        body["url"] = url;

        return await _http.PostAsync<CrawlResponse>(
            "/v2/crawl", body, cancellationToken: cancellationToken);
    }

    /// <summary>
    /// Gets the status and results of a crawl job.
    /// </summary>
    public async Task<CrawlJob> GetCrawlStatusAsync(
        string jobId,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(jobId);

        return await _http.GetAsync<CrawlJob>(
            $"/v2/crawl/{jobId}", cancellationToken);
    }

    /// <summary>
    /// Crawls a website and waits for completion (auto-polling).
    /// </summary>
    public async Task<CrawlJob> CrawlAsync(
        string url,
        CrawlOptions? options = null,
        int pollIntervalSec = DefaultPollIntervalSec,
        int timeoutSec = DefaultJobTimeoutSec,
        CancellationToken cancellationToken = default)
    {
        var start = await StartCrawlAsync(url, options, cancellationToken);
        return await PollCrawlAsync(
            start.Id ?? throw new FirecrawlException("Crawl start did not return a job ID"),
            pollIntervalSec, timeoutSec, cancellationToken);
    }

    /// <summary>
    /// Cancels a running crawl job.
    /// </summary>
    public async Task<Dictionary<string, object>> CancelCrawlAsync(
        string jobId,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(jobId);

        return await _http.DeleteAsync<Dictionary<string, object>>(
            $"/v2/crawl/{jobId}", cancellationToken);
    }

    /// <summary>
    /// Gets errors from a crawl job.
    /// </summary>
    public async Task<Dictionary<string, object>> GetCrawlErrorsAsync(
        string jobId,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(jobId);

        return await _http.GetAsync<Dictionary<string, object>>(
            $"/v2/crawl/{jobId}/errors", cancellationToken);
    }

    // ================================================================
    // BATCH SCRAPE
    // ================================================================

    /// <summary>
    /// Starts an async batch scrape job.
    /// </summary>
    public async Task<BatchScrapeResponse> StartBatchScrapeAsync(
        List<string> urls,
        BatchScrapeOptions? options = null,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(urls);

        var body = BuildBody(options);
        body["urls"] = urls;

        // The API expects scrape options flattened at the top level
        if (body.TryGetValue("options", out var nested) && nested is JsonElement nestedElement)
        {
            body.Remove("options");
            var nestedDict = JsonSerializer.Deserialize<Dictionary<string, object>>(
                nestedElement.GetRawText(), FirecrawlHttpClient.JsonOptions);
            if (nestedDict != null)
            {
                var batchFields = new Dictionary<string, object>(body);
                foreach (var kv in nestedDict)
                    body.TryAdd(kv.Key, kv.Value);
                foreach (var kv in batchFields)
                    body[kv.Key] = kv.Value;
            }
        }

        Dictionary<string, string>? extraHeaders = null;
        if (options?.IdempotencyKey is { Length: > 0 } idempotencyKey)
        {
            extraHeaders = new Dictionary<string, string>
            {
                ["x-idempotency-key"] = idempotencyKey
            };
        }

        return await _http.PostAsync<BatchScrapeResponse>(
            "/v2/batch/scrape", body, extraHeaders, cancellationToken);
    }

    /// <summary>
    /// Gets the status and results of a batch scrape job.
    /// </summary>
    public async Task<BatchScrapeJob> GetBatchScrapeStatusAsync(
        string jobId,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(jobId);

        return await _http.GetAsync<BatchScrapeJob>(
            $"/v2/batch/scrape/{jobId}", cancellationToken);
    }

    /// <summary>
    /// Batch-scrapes URLs and waits for completion (auto-polling).
    /// </summary>
    public async Task<BatchScrapeJob> BatchScrapeAsync(
        List<string> urls,
        BatchScrapeOptions? options = null,
        int pollIntervalSec = DefaultPollIntervalSec,
        int timeoutSec = DefaultJobTimeoutSec,
        CancellationToken cancellationToken = default)
    {
        var start = await StartBatchScrapeAsync(urls, options, cancellationToken);
        return await PollBatchScrapeAsync(
            start.Id ?? throw new FirecrawlException("Batch scrape start did not return a job ID"),
            pollIntervalSec, timeoutSec, cancellationToken);
    }

    /// <summary>
    /// Cancels a running batch scrape job.
    /// </summary>
    public async Task<Dictionary<string, object>> CancelBatchScrapeAsync(
        string jobId,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(jobId);

        return await _http.DeleteAsync<Dictionary<string, object>>(
            $"/v2/batch/scrape/{jobId}", cancellationToken);
    }

    // ================================================================
    // MAP
    // ================================================================

    /// <summary>
    /// Discovers URLs on a website.
    /// </summary>
    public async Task<MapData> MapAsync(
        string url,
        MapOptions? options = null,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(url);

        var body = BuildBody(options);
        body["url"] = url;

        var response = await _http.PostAsync<ApiResponse<MapData>>(
            "/v2/map", body, cancellationToken: cancellationToken);

        return response.Data ?? throw new FirecrawlException("Map response contained no data");
    }

    // ================================================================
    // SEARCH
    // ================================================================

    /// <summary>
    /// Performs a web search.
    /// </summary>
    public async Task<SearchData> SearchAsync(
        string query,
        SearchOptions? options = null,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(query);

        var body = BuildBody(options);
        body["query"] = query;

        var response = await _http.PostAsync<ApiResponse<SearchData>>(
            "/v2/search", body, cancellationToken: cancellationToken);

        return response.Data ?? throw new FirecrawlException("Search response contained no data");
    }

    // ================================================================
    // USAGE & METRICS
    // ================================================================

    /// <summary>
    /// Gets current concurrency usage.
    /// </summary>
    public async Task<ConcurrencyCheck> GetConcurrencyAsync(
        CancellationToken cancellationToken = default)
    {
        return await _http.GetAsync<ConcurrencyCheck>(
            "/v2/concurrency-check", cancellationToken);
    }

    /// <summary>
    /// Gets current credit usage.
    /// </summary>
    public async Task<CreditUsage> GetCreditUsageAsync(
        CancellationToken cancellationToken = default)
    {
        return await _http.GetAsync<CreditUsage>(
            "/v2/team/credit-usage", cancellationToken);
    }

    // ================================================================
    // INTERNAL POLLING HELPERS
    // ================================================================

    private async Task<CrawlJob> PollCrawlAsync(
        string jobId,
        int pollIntervalSec,
        int timeoutSec,
        CancellationToken cancellationToken)
    {
        var deadline = DateTime.UtcNow.AddSeconds(timeoutSec);

        while (DateTime.UtcNow < deadline)
        {
            cancellationToken.ThrowIfCancellationRequested();

            var job = await GetCrawlStatusAsync(jobId, cancellationToken);
            if (job.IsDone)
                return await PaginateCrawlAsync(job, cancellationToken);

            await Task.Delay(TimeSpan.FromSeconds(pollIntervalSec), cancellationToken);
        }

        throw new JobTimeoutException(jobId, timeoutSec, "Crawl");
    }

    private async Task<BatchScrapeJob> PollBatchScrapeAsync(
        string jobId,
        int pollIntervalSec,
        int timeoutSec,
        CancellationToken cancellationToken)
    {
        var deadline = DateTime.UtcNow.AddSeconds(timeoutSec);

        while (DateTime.UtcNow < deadline)
        {
            cancellationToken.ThrowIfCancellationRequested();

            var job = await GetBatchScrapeStatusAsync(jobId, cancellationToken);
            if (job.IsDone)
                return await PaginateBatchScrapeAsync(job, cancellationToken);

            await Task.Delay(TimeSpan.FromSeconds(pollIntervalSec), cancellationToken);
        }

        throw new JobTimeoutException(jobId, timeoutSec, "Batch scrape");
    }

    private async Task<CrawlJob> PaginateCrawlAsync(
        CrawlJob job,
        CancellationToken cancellationToken)
    {
        job.Data ??= new List<Document>();
        var current = job;

        while (!string.IsNullOrEmpty(current.Next))
        {
            var nextPage = await _http.GetAbsoluteAsync<CrawlJob>(
                current.Next, cancellationToken);

            if (nextPage.Data is { Count: > 0 })
                job.Data.AddRange(nextPage.Data);

            current = nextPage;
        }

        job.Next = null;
        return job;
    }

    private async Task<BatchScrapeJob> PaginateBatchScrapeAsync(
        BatchScrapeJob job,
        CancellationToken cancellationToken)
    {
        job.Data ??= new List<Document>();
        var current = job;

        while (!string.IsNullOrEmpty(current.Next))
        {
            var nextPage = await _http.GetAbsoluteAsync<BatchScrapeJob>(
                current.Next, cancellationToken);

            if (nextPage.Data is { Count: > 0 })
                job.Data.AddRange(nextPage.Data);

            current = nextPage;
        }

        job.Next = null;
        return job;
    }

    // ================================================================
    // INTERNAL UTILITIES
    // ================================================================

    private static Dictionary<string, object> BuildBody(object? options)
    {
        if (options == null)
            return new Dictionary<string, object>();

        var json = JsonSerializer.Serialize(options, FirecrawlHttpClient.JsonOptions);
        return JsonSerializer.Deserialize<Dictionary<string, object>>(json, FirecrawlHttpClient.JsonOptions)
            ?? new Dictionary<string, object>();
    }

    private static string ResolveApiKey(string? apiKey)
    {
        if (!string.IsNullOrWhiteSpace(apiKey))
            return apiKey;

        var envKey = Environment.GetEnvironmentVariable("FIRECRAWL_API_KEY");
        if (!string.IsNullOrWhiteSpace(envKey))
            return envKey;

        throw new FirecrawlException(
            "API key is required. Pass it to the constructor or set the FIRECRAWL_API_KEY environment variable.");
    }

    private static string ResolveApiUrl(string? apiUrl)
    {
        if (!string.IsNullOrWhiteSpace(apiUrl))
            return apiUrl;

        var envUrl = Environment.GetEnvironmentVariable("FIRECRAWL_API_URL");
        if (!string.IsNullOrWhiteSpace(envUrl))
            return envUrl;

        return DefaultApiUrl;
    }
}
